### PR TITLE
	SonarQube analysis on Maven and Gradle fails if the SQ server is over https

### PR DESCRIPTION
**Pull request by bgavrilMS**
Replace the use of the low-level http and https modules with a higher level request module.

### CODE REVIEW

File: Tasks/Gradle/CodeAnalysis/SonarQube/endpoint.ts [hunk 1]
```diff
@@ -15,6 +15,8 @@ export class SonarQubeEndpoint {
         var genericEndpointName: string = tl.getInput('sqConnectedServiceName');
         var hostUrl = tl.getEndpointUrl(genericEndpointName, false);
 
+        tl.debug(`[SQ] SonarQube endpoint: ${hostUrl}`);
```
Comment (bgavrilMS) [inline Tasks/Gradle/CodeAnalysis/SonarQube/endpoint.ts]: Unrelated change - I added more logging

File: Tasks/Gradle/CodeAnalysis/SonarQube/server.ts [hunk 2]
```diff
@@ -1,9 +1,5 @@
 import Q = require('q');
-import url = require('url');
-import https = require('https');
-import http = require('http');
-import {IncomingMessage} from 'http';
-
+import request = require('request');
```
Comment (bgavrilMS) [inline Tasks/Gradle/CodeAnalysis/SonarQube/server.ts]: LCA approved

File: Tasks/Gradle/CodeAnalysis/SonarQube/server.ts [hunk 3]
```diff
@@ -12,94 +8,58 @@ export interface ISonarQubeServer {
...
     }
 
-    public invokeApiCall(path:string):Q.Promise<Object> {
-        var defer = Q.defer<Object>();
-
-        var options:any = this.createSonarQubeHttpRequestOptions(path);
-
-        // Dynamic switching - we cannot use https.request() for http:// calls or vice versa
-        var protocolToUse;
-        switch (options.protocol) {
```
Comment (bgavrilMS) [inline Tasks/Gradle/CodeAnalysis/SonarQube/server.ts]: manually switching between http and https proves unreliable

File: definitions/form-data.d.ts [hunk 4]
```diff
@@ -0,0 +1,25 @@
+// Type definitions for form-data
```
Comment (bgavrilMS) [inline definitions/form-data.d.ts]: This type definition is required by the "request" type def

File: Tasks/Gradle/CodeAnalysis/SonarQube/server.ts [hunk 5]
```diff
@@ -12,94 +8,58 @@ export interface ISonarQubeServer {
...
-            host: hostUrl.hostname,
-            port: hostUrl.port || 80,
-            path: path || '/',
-            auth: authUser + ':' + authPass,
-            headers: {}
-        };
+            baseUrl: this.endpoint.Url,
+            uri: path,
+            json: true,
+            'auth': {
```
Comment (bgavrilMS) [inline Tasks/Gradle/CodeAnalysis/SonarQube/server.ts]: I tested: annon auth + https, basic auth with username and password + http, basic auth with token + http

File: definitions/form-data.d.ts (hunk 4 above)
Comment (duncanpMS) [inline definitions/form-data.d.ts]: Does the use of the type definitions require LCA approval too?

**Comment by mjcheetham** [comment]
LGTM

File: definitions/form-data.d.ts (hunk 4 above)
Comment (bgavrilMS) [inline definitions/form-data.d.ts]: No, as the d.ts files are not distributed with the tasks.
